tonic/src/transport/endpoint.rs

```diff
@@ -252,3 +203,96 @@ impl fmt::Debug for Endpoint {
         f.debug_struct("Endpoint").finish()
     }
 }
+
+/// Configures TLS settings for endpoints.
+#[cfg(feature = "tls")]
+pub struct ClientTlsConfig {
+    provider: TlsProvider,
+    domain: Option<String>,
+    cert: Option<Certificate>,
+    #[cfg(feature = "openssl")]
+    openssl_raw: Option<openssl1::ssl::SslConnector>,
+    #[cfg(feature = "rustls")]
+    rustls_raw: Option<tokio_rustls::rustls::ClientConfig>,
+}
+
+#[cfg(feature = "tls")]
+impl fmt::Debug for ClientTlsConfig {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ClientTlsConfig")
+            .field("provider", &self.provider)
+            .finish()
+    }
+}
+
+#[cfg(feature = "tls")]
+impl ClientTlsConfig {
+    /// Creates a new `ClientTlsConfig` backed by the specified provider. Enable the `openssl` or
+    /// `rustls` features of the `tonic` crate to use OpenSSL or Rustls respectively.
+    pub fn new(provider: TlsProvider) -> Self {
+        ClientTlsConfig {
+            provider,
+            domain: None,
+            cert: None,
+            #[cfg(feature = "openssl")]
+            openssl_raw: None,
+            #[cfg(feature = "rustls")]
+            rustls_raw: None,
+        }
+    }
+
+    /// Sets the domain name against which to verify the server's TLS certificate. If set to `None`
+    /// (the default), the address specified
+    pub fn domain_name(mut self, domain_name: impl Into<Option<String>>) -> Self {
+        self.domain = domain_name.into();
+        self
+    }
+
+    /// Sets the CA Certificate against which to verify the server's TLS certificate.
+    pub fn ca_certificate(mut self, ca_certificate: Certificate) -> Self {
+        self.cert = Some(ca_certificate);
+        self
+    }
+
+    /// Use options specified by the given `SslConnector` to configure TLS.
+    ///
+    /// This overrides all other TLS options set via other means.
+    #[cfg(feature = "openssl")]
+    pub fn openssl_connector(mut self, connector: openssl1::ssl::SslConnector) -> Self {
+        self.openssl_raw = Some(connector);
+        self
+    }
+
+    /// Use options specified by the given `ClientConfig` to configure TLS.
+    ///
+    /// This overrides all other TLS options set via other means.
+    #[cfg(feature = "rustls")]
+    pub fn rustls_client_config(mut self, config: tokio_rustls::rustls::ClientConfig) -> Self {
+        self.rustls_raw = Some(config);
+        self
+    }
+
+    fn tls_connector(self, uri: Uri) -> Option<TlsConnector> {
+        match self.provider {
+            TlsProvider::None => None,
+            #[cfg(feature = "openssl")]
+            TlsProvider::OpenSsl => {
+                if let Some(connector) = self.openssl_raw {
+                    return Some(TlsConnector::new_with_openssl_raw(connector).unwrap());
```

Comment:
do we want to bubble this result up? or is this never going to panic?

Reply:
I've made this bubble up to the top level of the builder, but we should probably have a wider think about how errors are handled during construction?

Reply:
Yeah, I agree, I have some other thoughts about the builders I'd like to change. For now I think the best thing is to at least bubble it up and let the user decide if they want to panic or not.

Reply:
Right now I've adjusted it such that the various builders return a `Result<Tls{Connector,Acceptor}, crate::Error>`, and the unwrap happens in the `tls_config()` method on the builder - in the same way as it was before. Not sure I see a good way to propagate errors in those builders without some kind of change to the builder pattern as you allude to there.

Reply:
You're right! I totally unwrapped before, I think this is fine then. Sorry about that. 😅